src/jdk.incubator.vector/share/classes/jdk/incubator/vector/VectorMathLibrary.java

```diff
@@ -0,0 +1,311 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.incubator.vector;
+
+import jdk.internal.util.StaticProperty;
+import jdk.internal.vm.annotation.DontInline;
+import jdk.internal.vm.annotation.ForceInline;
+import jdk.internal.vm.annotation.Stable;
+import jdk.internal.vm.vector.VectorSupport;
+
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.SymbolLookup;
+import java.util.function.IntFunction;
+
+import static jdk.incubator.vector.Util.requires;
+import static jdk.incubator.vector.VectorOperators.*;
+import static jdk.internal.util.Architecture.isX64;
+import static jdk.internal.vm.vector.Utils.debug;
+
+/**
+ * A wrapper for native vector math libraries bundled with the JDK (SVML and SLEEF).
+ * Binds vector operations to native implementations provided by the libraries.
+ */
+/*package-private*/ class VectorMathLibrary {
+    private static final SymbolLookup LOOKUP = SymbolLookup.loaderLookup();
+
+    interface Library {
+        String symbolName(Operator op, VectorSpecies<?> vspecies);
+        boolean isSupported(Operator op, VectorSpecies<?> vspecies);
+
+        String SVML  = "svml";
+        String SLEEF = "sleef";
+        String JAVA  = "java";
+
+        static Library getInstance() {
+            String libraryName = System.getProperty("jdk.incubator.vector.VectorMathLibrary", getDefaultName());
+            try {
+                return switch (libraryName) {
+                    case SVML  -> new SVML();
+                    case SLEEF -> new SLEEF();
+                    case JAVA  -> new Java();
+                    default    -> throw new IllegalArgumentException("Unsupported vector math library: " + libraryName);
+                };
+            } catch (Throwable e) {
+                debug("Error during initialization of %s library: %s", libraryName, e);
+                return new Java(); // fallback
+            }
+        }
+
+        static String getDefaultName() {
+            return switch (StaticProperty.osArch()) {
+                case "amd64", "x86_64" -> SVML;
+                case "aarch64" -> SLEEF;
+                default -> JAVA;
+            };
+        }
+    }
+
+    private static final Library LIBRARY = Library.getInstance();
+
+    static {
+        debug("%s library is used (cpu features: %s)", LIBRARY.getClass().getSimpleName(), CPUFeatures.features());
+    }
+
+    private static class Java implements Library {
+        @Override
+        public String symbolName(Operator op, VectorSpecies<?> vspecies) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean isSupported(Operator op, VectorSpecies<?> vspecies) {
+            return false; // always use default implementation
+        }
+    }
+
+    /**
+     * Naming convention in SVML vector math library.
```

Comment:
Does this library has code for all AVX configurations?

Reply:
Yes, there are 4 configurations (`-XX:UseAVX=[0..3]`) in total covered by SVML library.

Reply:
Good.